pkg/monitor/corednsmonitor.go

```diff
@@ -42,24 +43,34 @@ func CorednsWatch(kubeconfigPath, clusterConfigPath, templatePath, cfgPath strin
 			if err != nil {
 				return err
 			}
```

Comment:
I checked deployment with both MCO and baremetal-runtimecfg PRs,  Corefile includes the nodes records.

I have two comments:

A. It could be useful to add log message with details about the cause for Corefile rendering.
B. Maybe it worth to add stabilization counter for changes before updating the Corefile, I checked coredns-monitor logs in my env [1] and it seems that Corefile was rendered 8 times .


[1] 

[core@master-1 ~]$ sudo crictl logs ac1436d7e367a
time="2021-03-18T11:17:35Z" level=error msg="Failed to build client config: invalid configuration: [unable to read client-cert /var/lib/kubelet/pki/kubelet-client-current.pem for default-auth due to open /var/lib/kubelet/pki/kubelet-client-current.pem: no such file or directory, unable to read client-key /var/lib/kubelet/pki/kubelet-client-current.pem for default-auth due to open /var/lib/kubelet/pki/kubelet-client-current.pem: no such file or directory]"
time="2021-03-18T11:17:35Z" level=error msg="Failed to build client config: invalid configuration: [unable to read client-cert /var/lib/kubelet/pki/kubelet-client-current.pem for default-auth due to open /var/lib/kubelet/pki/kubelet-client-current.pem: no such file or directory, unable to read client-key /var/lib/kubelet/pki/kubelet-client-current.pem for default-auth due to open /var/lib/kubelet/pki/kubelet-client-current.pem: no such file or directory]"
time="2021-03-18T11:18:05Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:18:35Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:20:43Z" level=error msg="Failed to get node list: etcdserver: request timed out"
time="2021-03-18T11:20:43Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:21:13Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:29:09Z" level=error msg="Failed to get node list: the server was unable to return a response in the time allotted, but may still be processing the request (get nodes)"
time="2021-03-18T11:29:09Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:29:39Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:35:39Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
time="2021-03-18T11:36:09Z" level=info msg="Runtimecfg rendering template" path=/etc/coredns/Corefile
[core@master-1 ~]$

Reply:
> A. It could be useful to add log message with details about the cause for Corefile rendering.

Sure, I can do that.

> B. Maybe it worth to add stabilization counter for changes before updating the Corefile, I checked coredns-monitor logs in my env [1] and it seems that Corefile was rendered 8 times .

Ah, interesting. I think this is actually a bug. When it failed to reach the api server it re-rendered the config, which shouldn't happen.

The monitor only runs every 30 seconds, so I don't think it's likely to do much flapping. In a normal deployment I would expect up to 5 re-renders (assuming our default 5 node deployment) as each node becomes available. I would prefer to make sure the node list is as up to date as possible rather than try to limit the number of times we render the config. It's not that resource intensive and there's no outage (unlike keepalived and haproxy) when we do it.

I do need to fix the error handling so we aren't rendering empty node lists though.

Reply:
> Ah, interesting. I think this is actually a bug. When it failed to reach the api server it re-rendered the config, which shouldn't happen.

Does that mean that in such cases we'll keep the previously rendered nodes if they exist? Does this have any effect on power outages?

Reply:
Yep, it looks better with the latest commit, the nodes entries were added to coredns Corefile and coredns file was rendered twice.

Reply:
> Does that mean that in such cases we'll keep the previously rendered nodes if they exist? Does this have any effect on power outages?

Yes, if we lose contact with the API then we'll leave the node list as it was the last time we were able to update. I would expect that to be better for recovery scenarios. Once the API comes back up we'll be able to update, but in the meantime all of the previous DNS records will still be available. This should better match what happens on other platforms where the DNS records come from the cloud DNS instead of in-cluster.